### PR TITLE
Adding capability of taking auxiliary data

### PR DESCRIPTION
**Pull request by yinweisu**
The idea of this PR is to enable benchmark yaml taking in auxiliary data. Auxiliary data could be useful when the task requires more than train and test datasets. For example, images are needed for multimodality tasks. There are more usage possibles, and therefore we think it's worth to have auxiliary data as a general feature.

Example yaml file:
```
- name: benchmark
  folds: 1
  dataset:
    target: label
    train: train.csv
    test: test.csv
  auxiliary_data:
    train: train_aux.zip
    test: test_aux.zip  # test aux not required
```

The main difference between auxiliary data and regular dataset should be that:
1. We don't require a 1 to 1 correspondence of train and test aux data. Imagine a use case where some psudo label data are used during training, while we don't need those during testing.
2. AMLB should let the users to handle the aux data themselves because use case of the aux data could vary a lot. AMLB only prepare the data and hand users the path.

### CODE REVIEW

**Comment by yinweisu** [comment]
There are duplicate code in terms of extracting auxiliary paths because I find it confusing to put the logic of extracting auxiliary and regular train/test paths in a single function. I'm sure there are better designs and feel free to propose them :).

**Comment by yinweisu** [comment]
@sebhrusen @PGijsbers

File: amlb/datasets/file.py [hunk 1]
```diff
@@ -167,6 +245,59 @@ def _get_metadata(self, prop):
         return meta[prop]
 
 
+class DatasetWithauxiliaryData:
```
Comment (Innixma) [inline amlb/datasets/file.py:248]: DatasetWithAuxiliaryData (Proper capitalization)

**Comment by PGijsbers** [comment]
Thank you for the contribution. I just wanted to let you know that it might be a little while before I myself can look at this. Though if Seb finds the time then I'm okay with whatever he says :)

In general I think it would be a useful addition to allow for auxiliary data to be present in a task. These types of tasks have already been the focus of research and AutoML competitions. Not processing auxiliary data in any meaningful way by the benchmark framework is the only way I see it work (given its free form).

File: amlb/datasets/file.py [hunk 2]
```diff
@@ -167,6 +245,59 @@ def _get_metadata(self, prop):
         return meta[prop]
 
 
+class DatasetWithAuxiliaryData:
```
Comment (sebhrusen) [inline amlb/datasets/file.py:248]: must extend `Dataset` if you want to expose this.

Actually, I think the right interface would be sth like:
```python
# in data.py

class Datasplit(ABC):
    ...

    @property
    def has_auxiliary_data(self) -> bool:
        pass
       
    @property
    def auxiliary_data(self) -> List[AuxData]:
        pass
        
class AuxData(ABC):
    # does it need a name property? a type/category/labels?

    @property
    def path(self) -> str:
        pass

    @property
    @abstractmethod
    def data(self) -> DF:
        """
        :return: the auxiliary data as a pandas DataFrame.
        """
        pass

    @profile(logger=log)
    def release(self, properties=None):
        clear_cache(self, properties)
```
then you can implement the logic for files to fulfill the contract of this interface.

File: amlb/datasets/file.py [hunk 3]
```diff
@@ -55,6 +55,84 @@ def load(self, dataset, fold=0):
...
 
+    @profile(logger=log)
+    def load_auxiliary_data(self, auxiliary_data, fold=0):
+        auxiliary_data = auxiliary_data if isinstance(auxiliary_data, ns) else ns(path=auxiliary_data)
+        log.debug("Loading auxiliary data %s", auxiliary_data)
+        paths = self._extract_auxiliary_paths(auxiliary_data.path if 'path' in auxiliary_data else auxiliary_data, fold=fold)
+        train_path = paths['train'][fold]
+        test_path = paths['test'][fold]
+        paths = dict(train=train_path, test=test_path)
+        return paths
```
[the remark this answers is not in the document]
Comment (yinweisu) [reply amlb/datasets/file.py:66]: The problem here is that the auxiliary data could be in all different forms. For example, it could be a zip file containing bunch of images. We don't know how should we handle it

[the remark this answers is not in the document]
Comment (sebhrusen) [reply amlb/datasets/file.py:66]: I understand, see my suggestion for `AuxData` interface below: we can easily imagine extending it with some `format` property later.
Right now, I'm just concerned about providing an access to those aux data that is likely to be useful only for AG, that's why I just want to have a better interface.

[the remark this answers is not in the document]
Comment (yinweisu) [reply amlb/datasets/file.py:66]: So in the `AuxData` interface, there's a `data` property that's supposed to return a DF, which is not doable for many aux data. My point of view is that each framework should handle their own usage of the aux data because of its free form. I don't think providing only the path is useful only for AG.

File: amlb/datasets/file.py [hunk 4]
```diff
@@ -55,6 +55,84 @@ def load(self, dataset, fold=0):
...
+
+    def _extract_auxiliary_paths(self, auxiliary_data, fold=None):
```
[the remark this answers is not in the document]
Comment (yinweisu) [reply amlb/datasets/file.py:68]: There are actually some other minor changes:
1. we don't unzip auxiliary data because of its free form.
2. we don't require users providing both train and test auxiliary data.

These would require a flag checking whether we are extracting auxiliary data, and would make it less general.

File: amlb/datasets/file.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sebhrusen) [reply amlb/datasets/file.py:66]: > So in the AuxData interface, there's a data property that's supposed to return a DF, which is not doable for many aux data

hence the idea of adding a type/format attribute to `AuxData`.
Again, you don't have to implement the `data` property now, I just want to have this `AuxData` interface now, so that we can implement more specific loading of those data later without having to rewrite everything.

[the remark this answers is not in the document]
Comment (yinweisu) [reply amlb/datasets/file.py:66]: This makes sense. Thanks for the discussion!

File: amlb/datasets/file.py (hunk 2 above)
Comment (sebhrusen) [reply amlb/datasets/file.py]: > Having auxiliary data in Datasplit seems unnecessary

why? you were planning to have `dataset.train_auxiliary_path` and `dataset.test_auxiliary_path`.
I'm suggesting that for you it would then be accessed as `dataset.train.auxiliary_data[0].path` and `dataset.test.auxiliary_data[0].path`: it's much more consistent with the current interfaces.

Everything that is exposed to frameworks must be defined in one of the interfaces defined in `data.py`.

> This is identical to DatasetWithAuxiliaryData

no, this is not identical: you're creating a class that doesn't extend anything and you add features that are not visible in the contract defined by the interfaces: those are the only references for all data providers.

to sum up, a dataset has currently 2 splits: `train` and `test`.
you can have aux data for each of those, therefore auxdata should be a property of `Datasplit`: if you add the auxdata directly to the Dataset instance instead of Datasplit, you create an inconsistency.

Comment (yinweisu) [reply amlb/datasets/file.py]: Makes sense. So do you think adding method to Dataset to add aux data to existing data split is a good idea? Or we would need to have loading of dataset and aux data at the same time then construct the Dataset object. This would require refactor of DataLoader.load

**Comment by yinweisu** [comment]
@sebhrusen Hey Seb, happy holidays! Can you review the code when you have time?

File: amlb/datasets/file.py [hunk 5]
```diff
@@ -55,6 +55,84 @@ def load(self, dataset, fold=0):
         else:
             raise ValueError(f"Unsupported file type: {ext}")
 
+    @profile(logger=log)
+    def load_auxiliary_data(self, auxiliary_data, fold=0):
```
[the remark this answers is not in the document]
Comment (yinweisu) [reply amlb/datasets/file.py]: Added

**Comment by sebhrusen** [comment]
Hey @yinweisu sorry for making you wait on this: don't have much time for `amlb` right now, but I'll try to look at your PR during the week.
Also, I'll try to work soon with @PGijsbers on restructuring some parts of the code and setting up a workflow for contributors to make this easier for you to contribute and for us to review without being afraid of breaking existing logic.
Thanks for your understanding.

**Comment by Innixma** [comment]
We have an intern joining for AutoGluon that we want to have work on multi-modal optimization (i.e. datasets with an image feature). Ideally we would like him to be able to use AutoMLBenchmark as the benchmarking tool, but this is tricky without this functionality being merged in. We can make do by hacking it into a forked repo, but just mentioning this.
